bip-csfs.md

```diff
@@ -0,0 +1,133 @@
+```
+BIP: XXX
+Layer: Consensus (soft fork)
+Title: CHECKSIGFROMSTACK
+Author: Brandon Black <freedom@reardencode.com>, Jeremy Rubin <j@rubin.io>
+Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-XXXX
+Status: Draft
+Type: Standards Track
+Created: 2023-12-22
+License: BSD-3-CLAUSE
+```
+
+## Abstract
+
+This BIP describes two new opcodes for the purpose of checking cryptographic
+signatures in bitcoin scripts against data other than bitcoin transactions.
+
+## Summary
+
+When verifying taproot script spends having leaf version 0xc0 (as defined in
+[BIP 342]), we propose `OP_CHECKSIGFROMSTACK` to replace `OP_SUCCESS204`
+(0xcc).
+
+`OP_CHECKSIGFROMSTACKVERIFY` has semantics similar to `OP_CHECKSIG`, as
+specified below.
+
+Only 32-byte keys are constrained. Similar to [BIP 341] unknown key types, for
+other key lengths no signature verification is performed.
+
+## Specification
+
+* If fewer than 3 elements are on the stack, the script MUST fail and terminate immediately.
+* The public key (top element), message (second to top element), and signature (third from top element) are read from the stack.
+* The top three elements are popped from the stack.
+* If the public key size is zero, the script MUST fail and terminate immediately.
+* If the public key size is 32 bytes, it is considered to be a public key as described in [BIP 340]:
+    * If the signature is not the empty vector, the signature is validated against the public key and message according to [BIP 340]. Validation failure in this case immediately terminates script execution with failure.
+* If the public key size is not zero, and it is not a [BIP 340] public key; the public key is of an unknown public key type, and no actual signature verification is applied. During script execution of signature opcodes they behave exactly as known public key types except that signature validation is considered to be successful.
```

Comment:
I’m not sure whether this is intentional, but previously we treat a 32-byte key as a BIP 340 public key. I assume if it were out-of-range, the signature validation would fail.
However here, rather than repeating that the lengths 0-byte and 32-byte are excluded, the criterion appears to be "not zero and not a BIP 340 public key". I assume this is covered by the previous line, but what if a 32-byte value were provided that is out of range, would this line apply or not? 

Is it possible that this should read:

```suggestion
* If the public key size is not zero bytes and not 32 bytes; the public key is of an unknown public key type. Signature verification for unknown public key types succeeds as if signature verification for a known public key type had succeeded.
```

Reply:
I believe I copied that text from BIP340. Your wording is more clear.